Add Sauce Labs support to wptrunner

bobholt opened the pull request:
This PR builds on @jgraham's original [sauce labs branch](https://github.com/w3c/wptrunner/compare/jgraham/sauce) as [modified by](https://github.com/w3c/wptrunner/compare/master...jeffcarp:sauce) @jeffcarp.

You can see this working (in conjunction with changes to web-platform-test and prbuildbot) here: https://github.com/bobholt/web-platform-tests/pull/10

In contrast to the previous work, this uses the same selenium executor as the other tests. This allows the test results to be run and reported back in the same way. This is made possible by disabling the pop-up blockers on Edge and Safari with Sauce Labs pre-run scripts. The previous work got around this by running the tests on w3c-test.org and scraping the output HTML.

Note that a piece of housekeeping will be to move the pre-run scripts referenced in `wptrunner/browsers/sauce.py` to a permanent location (currently in place as gists on my account). They can be separately on w3c-test.org or included in this repo (and linked from the raw master).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/wptrunner/243)
<!-- Reviewable:end -->

jgraham commented:
Reviewed 2 of 13 files at r1, 15 of 19 files at r2.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 8 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJ9tYoEhZGg3hn43Pr:-KgJ9tYpZwHzwQfo_aKy:b1b2vix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/.gitignore#L8)):*
> ```
> test/test.cfg
> test/metadata/MANIFEST.json
> tools
> ```

That seems wrong; we want to know when we are on the wrong submodule revision.

---

*[wptrunner/environment.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJA7tTurURWqkYKCts:-KgJA7tU0leEZQZ-FoJ-:b-wo8gc0) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L31)):*
> ```Python
> 
> host = "web-platform.test"
> hostnames = ["%s.%s" % (item, host) if item else host for item in subdomains]
> ```

Any specific reason for this change? I need to add a second host to wpt soon, so this won't work as well in that case.

---

*[wptrunner/environment.py, line 101 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBLxzXcCn9CtOs8fW:-KgJBLxzXcCn9CtOs8fX:b-1cr4we) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L101)):*
> ```Python
>         self.cache_manager = multiprocessing.Manager()
>         self.stash = serve.stash.StashServer()
>         self.routes = self.get_routes()
> ```

This now seemes to be called here and when starting the server below.

---

*[wptrunner/environment.py, line 117 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJAOZvE1gR0RC354Zr:-KgJAOZvE1gR0RC354Zs:b-835nra) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L117)):*
> ```Python
> 
>     def __exit__(self, exc_type, exc_val, exc_tb):
>         self.cache_manager.__exit__(exc_type, exc_val, exc_tb)
> ```

This is now duplicated here.

---

*[wptrunner/environment.py, line 222 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBmmzmpJTkzKzvR6E:-KgJBmmzmpJTkzKzvR6F:bsa4mk1) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L222)):*
> ```Python
>     def __init__(self, test_paths, ssl_env, pause_after_test, debug_info, options):
>         TestEnvironment.__init__(self, test_paths, ssl_env, pause_after_test, debug_info, options)
>         assert not self.pause_after_test
> ```

I think a better option might just be to ignore this argument if we use `--product=sauce`.

---

*[wptrunner/wptcommandline.py, line 193 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJVDaGb5JcH9-yf6Y2:-KgJVDaHNvSy06rRGqae:b-wlsm8h) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/wptcommandline.py#L193)):*
> ```Python
>     sauce_group.add_argument("--sauce-build", dest="sauce_build",
>                              help="Sauce Labs build identifier")
>     sauce_group.add_argument("--sauce-tag", dest="sauce_tag",
> ```

Should there just be one of these or should it take a list?

---

*[wptrunner/wptrunner.py, line 153 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJD46UgbjQptPaqywu:-KgJD46Vaa3T6ImAVama:bwndbgb) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/wptrunner.py#L153)):*
> ```Python
> 
>         with env_class(test_paths,
>                                  ssl_env,
> ```

Indentation is wrong.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgJOzxCCYbLciMEj8DN:bmmosw5) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
> ```Python
>     prerun_script = {
>         "MicrosoftEdge": {
>             "executable": "https://gist.githubusercontent.com/bobholt/4983aa95e65d55e7c1bf79c7b5d9c3af/raw/7c2d080a6da43c7f5a0b0190c1f0ed006935a0ce/edge-popup.bat",
> ```

These need to go somewhere in the wptrunner source. Have we checked in case `data` urls happen to work here?

---

*[wptrunner/browsers/sauce.py, line 39 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJPEGELqrF_kRTJS3D:-KgJPEGELqrF_kRTJS3E:b6rv3yz) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L39)):*
> ```Python
>         }
>     }
>     username = os.environ.get("SAUCE_USERNAME")
> ```

We should probably allow these to be set as command line options, maybe pointing to a file for the access key. At the very least we should fail in some understandable way if they aren't set.

---

*[wptrunner/browsers/sauce.py, line 52 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJZoB0fx_yoEnHHZML:-KgJZoB0fx_yoEnHHZMM:bwb3x5a) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L52)):*
> ```Python
>             "platform": platform,
>             "public": "public",
>             "tags": [tag, "CI"],
> ```

Why is CI always a tag? I might run this locally just to check some behaviour. What happens if we pass `[None, "CI"]`

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgJUuhsOGjVtfU1bH6L:b-4ds45b) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
> ```Python
>             "public": "public",
>             "tags": [tag, "CI"],
>             "tunnel-identifier": tunnel_id,
> ```

Does this rely on sauce-connect already running? Can we arrange to start it (and extract the tunnel_id? I didn't find this in the docs).

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgJS5WmA8A717OEhXo8:b-tgkizo) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
> ```Python
>     }
> 
>     if browser_name == 'safari':
> ```

Not for this PR, but I wonder if we should put the browser-specific parts into the relevant browser config file somehow.

Also, why is this needed?

---

*[wptrunner/browsers/sauce.py, line 102 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJbZvJUYdbn0A5eEkw:-KgJbZvJUYdbn0A5eEkx:b-i6rg2k) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L102)):*
> ```Python
> 
> 
>     # executor_kwargs["capabilities"] = get_capabilities(kwargs["sauce_config"])
> ```

Remove commented out code (and resulting dead code)

---

*[wptrunner/browsers/sauce.py, line 110 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJbf0xWIODi93l2HZl:-KgJbf0xWIODi93l2HZm:b-orkkbk) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L110)):*
> ```Python
>                                     kwargs["sauce_tunnel_id"])
> 
>     executor_kwargs["capabilities"] = sauce_config.get('capabilities')
> ```

So do we throw away everything apart from the capabilities? In that case why not just return the part we use?

---

*[wptrunner/browsers/sauce.py, line 118 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJcMHUzqCmiLTxo-lt:-KgJcMHUzqCmiLTxo-lu:bnxx8eu) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L118)):*
> ```Python
>     return {"host": "web-platform.test",
>             "bind_hostname": "true",
>             "supports_debugger": True}
> ```

It does?

---

*[wptrunner/executors/executorselenium.py, line 148 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJi8SZeu4vOl3e0CnK:-KgJi8S_9DZi7eebJeoN:b14l6ug) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/executorselenium.py#L148)):*
> ```Python
>         executor.start()
> 
>         flag = self.result_flag.wait(self.timeout + extra_timeout)
> ```

Why did this change? The intention is that there are three timeouts; the internal timeout, the webdriver script timeout and the final external timeout at which we just give up.

---

*[wptrunner/executors/testharness_seleniumremote.js, line 1 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJiWWoXr_Wvie9MhNj:-KgJiWWoXr_Wvie9MhNk:b-8giox3) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/testharness_seleniumremote.js#L1)):*
> ```JavaScript
> /* This Source Code Form is subject to the terms of the Mozilla Public
> ```

Is this actualy used anywhere now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[wptrunner/environment.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJA7tTurURWqkYKCts:-KgKA2x5bW7-X9DQKn1K:b-u5d0gp) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L31)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Any specific reason for this change? I need to add a second host to wpt soon, so this won't work as well in that case.
</blockquote></details>

This looks like a holdover from your original branch where the sauce job overrode the hostname (to use w3c-test.org). since that doesn't happen anymore, I'll put this back the way it was.

---

*[wptrunner/environment.py, line 222 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBmmzmpJTkzKzvR6E:-KgKJPZ1Ia4LEQiWnZ16:bn81wfx) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L222)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

I think a better option might just be to ignore this argument if we use `--product=sauce`.
</blockquote></details>

Since this eliminates the need for RemoteServerEnvironment, is there still some use in passing the `TestEnvironment` `env_class` explicitly in every browser file, or should this go back to implicitly using `TestEnvironment` for all browsers? Rephrased: is there a future where we would introduce another Environment class?

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgKPvo1AUqMkzjYlK-e:bfnqr4u) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

These need to go somewhere in the wptrunner source. Have we checked in case `data` urls happen to work here?
</blockquote></details>

I'll add these as files in the source. There's a (legacy?) `b2g_setup` directory in `/wptrunner/browsers`, so I'll put these in a `sauce_setup` directory unless there's a better suggestion. I don't see any docs supporting data-uris, but I can give it a try in a bit.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgKcblq2G-sFiYogbUb:b-pol0b6) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Does this rely on sauce-connect already running? Can we arrange to start it (and extract the tunnel_id? I didn't find this in the docs).
</blockquote></details>

This is an optional user-defined string. By defining it, it allows us to run jobs in parallel (Edge and Safari simultaneously). When using Travis's Sauce Connect configuration, this has to match the `TRAVIS_JOB_NUMBER` in order to match up with the correct VM. The `check_stability` script in my web-platform-tests PR defaults to that value when invoking wptrunner.

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgKeMFwNZPwrBDL1QUP:b-wmml7q) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Not for this PR, but I wonder if we should put the browser-specific parts into the relevant browser config file somehow.

Also, why is this needed?
</blockquote></details>

A bit more esoterica left out of the canonical Sauce docs: Safari 10 requires Apple's SafariDriver, not Selenium's. That requires Selenium running > v 3.0. However, Sauce defaults to Selenium 2.4.8 even when you select Safari 10 on macOs 10.12, so this has to be set explicitly. For a reason I have yet to determine, 3.3.0 breaks the Microsoft Edge run, so this conditional is here.

---

*[wptrunner/executors/executorselenium.py, line 148 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJi8SZeu4vOl3e0CnK:-KgKixAX34alqULmb2hA:bdtmd6t) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/executorselenium.py#L148)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Why did this change? The intention is that there are three timeouts; the internal timeout, the webdriver script timeout and the final external timeout at which we just give up.
</blockquote></details>

To be honest, I'm not quite sure what's going on with the timeouts. This was a change that was in your original work (https://github.com/w3c/wptrunner/compare/jgraham/sauce#diff-bd5bd3962137c5c4157e859c0681ed59R151) that I left in. I can put back the current logic.

---

*[wptrunner/executors/testharness_seleniumremote.js, line 1 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJiWWoXr_Wvie9MhNj:-KgKjp_BfUmVgNZHXbI5:b-1xw7it) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/testharness_seleniumremote.js#L1)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Is this actualy used anywhere now?
</blockquote></details>

Ah, nope!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 10 of 17 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 8 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJ9tYoEhZGg3hn43Pr:-KgKoSWSL7ibNauhJBw8:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/.gitignore#L8)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

That seems wrong; we want to know when we are on the wrong submodule revision.
</blockquote></details>

Done.

---

*[wptrunner/environment.py, line 101 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBLxzXcCn9CtOs8fW:-KgKnkVnTw9z5I5oNiPs:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L101)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

This now seemes to be called here and when starting the server below.
</blockquote></details>

Done.

---

*[wptrunner/environment.py, line 117 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJAOZvE1gR0RC354Zr:-KgKnkpFOV67Y4X4M4Qu:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L117)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

This is now duplicated here.
</blockquote></details>

Done.

---

*[wptrunner/wptcommandline.py, line 193 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJVDaGb5JcH9-yf6Y2:-KgKnmlydK-ty_P5qiEJ:b-5khs01) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/wptcommandline.py#L193)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Should there just be one of these or should it take a list?
</blockquote></details>

Definitely a list. Done.

---

*[wptrunner/wptrunner.py, line 153 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJD46UgbjQptPaqywu:-KgKnpE3-FcD0ANsYlmL:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/wptrunner.py#L153)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Indentation is wrong.
</blockquote></details>

Done.

---

*[wptrunner/browsers/sauce.py, line 39 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJPEGELqrF_kRTJS3D:-KgKnr1T25iqTjMvbVpC:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L39)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

We should probably allow these to be set as command line options, maybe pointing to a file for the access key. At the very least we should fail in some understandable way if they aren't set.
</blockquote></details>

Done.

---

*[wptrunner/browsers/sauce.py, line 52 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJZoB0fx_yoEnHHZML:-KgKnsUmzY-rAcW8LdlA:b-9ubso5) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L52)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Why is CI always a tag? I might run this locally just to check some behaviour. What happens if we pass `[None, "CI"]`
</blockquote></details>

Yeah, hardcoded just from getting it to work with Travis's recommendations. Above changes accept a list of tags and use that list instead. I don't see in the Sauce UI where these are used anyway.

---

*[wptrunner/browsers/sauce.py, line 102 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJbZvJUYdbn0A5eEkw:-KgKo54_DOP5Xg4t11cQ:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L102)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Remove commented out code (and resulting dead code)
</blockquote></details>

Done.

---

*[wptrunner/browsers/sauce.py, line 110 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJbf0xWIODi93l2HZl:-KgKo6F-H9MMnbnJ1P48:b-sqskcj) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L110)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

So do we throw away everything apart from the capabilities? In that case why not just return the part we use?
</blockquote></details>

Done. Refactored `get_capabilities` out and use that in the 2 places we need it. 

---

*[wptrunner/browsers/sauce.py, line 118 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJcMHUzqCmiLTxo-lt:-KgKoEjkYgxXE93K1h5F:bii09i5) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L118)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

It does?
</blockquote></details>

Done. Remnant of trying to get something useful out of Sauce. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Pushed a commit to take care of most of the requested changes. Testing on my fork. Will report back when it's confirmed working there.

bobholt commented:
This is back in working condition: https://github.com/bobholt/web-platform-tests/pull/14

jgraham commented:
Reviewed 8 of 9 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[wptrunner/environment.py, line 222 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBmmzmpJTkzKzvR6E:-KgL90AaHyuknX3Wy_Zj:baavp7w) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L222)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

Since this eliminates the need for RemoteServerEnvironment, is there still some use in passing the `TestEnvironment` `env_class` explicitly in every browser file, or should this go back to implicitly using `TestEnvironment` for all browsers? Rephrased: is there a future where we would introduce another Environment class?
</blockquote></details>

I think we shoud go back to a single environment and not add the complexity until we need it.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgLishXV-JiZ0xwff1z:b-d38v55) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

This is an optional user-defined string. By defining it, it allows us to run jobs in parallel (Edge and Safari simultaneously). When using Travis's Sauce Connect configuration, this has to match the `TRAVIS_JOB_NUMBER` in order to match up with the correct VM. The `check_stability` script in my web-platform-tests PR defaults to that value when invoking wptrunner.
</blockquote></details>

OK, so the question remains whether we can ensure that the sauce connect proxy is launched as part of the test run.

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgLI8tMoJR7jMkqNsSt:b-ecdyve) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

A bit more esoterica left out of the canonical Sauce docs: Safari 10 requires Apple's SafariDriver, not Selenium's. That requires Selenium running > v 3.0. However, Sauce defaults to Selenium 2.4.8 even when you select Safari 10 on macOs 10.12, so this has to be set explicitly. For a reason I have yet to determine, 3.3.0 breaks the Microsoft Edge run, so this conditional is here.
</blockquote></details>

Hmm, so if you want to use Firefox and geckodriver you need selenium 3.3.1. Is there a way to select latest selenium and set that for everything apart from Edge?

---

*[wptrunner/executors/executorselenium.py, line 148 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJi8SZeu4vOl3e0CnK:-KgLj1NRpPzdL86hr247:b26dzrq) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/executorselenium.py#L148)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

To be honest, I'm not quite sure what's going on with the timeouts. This was a change that was in your original work (https://github.com/w3c/wptrunner/compare/jgraham/sauce#diff-bd5bd3962137c5c4157e859c0681ed59R151) that I left in. I can put back the current logic.
</blockquote></details>

My original patch was largely written whilst waiting to go into the theatre for a recording of the BBC Radio comedy panel show Just A Minute. It is entirely possible it does insane things. I suggest we try reverting this and see if there was any reason for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgPKZeVTuOAS6KzCK0d:b4phstw) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

I'll add these as files in the source. There's a (legacy?) `b2g_setup` directory in `/wptrunner/browsers`, so I'll put these in a `sauce_setup` directory unless there's a better suggestion. I don't see any docs supporting data-uris, but I can give it a try in a bit.
</blockquote></details>

I will add that there is a "Sauce Storage" feature that would allow uploading these scripts to the Sauce virtual machine via the Sauce REST API at build time and (it seems) setting the executable URI to something like `"sauce-storage:edge-prerun.bat"`I was hesitant to do that at first because of the added complexity (just trying to get it to work at all) and the addition of another potential point of failure. However, this approach would avoid having to refer to a URL like https://raw.githubusercontent.com/w3c/wptrunner/045487371a47581e59feafbae0bea78210aa3369/wptrunner/browsers/sauce_setup/edge-prerun.bat

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgOzRENGwWYfZv_Oxgj:b6t1b2c) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

OK, so the question remains whether we can ensure that the sauce connect proxy is launched as part of the test run.
</blockquote></details>

Ah, I think I see what you're asking now. If I'm still being dense, feel free to talk it out with me on IRC.

I've done all of this through Travis, not setting up my own Sauce Connect locally. When we use this [through Travis](https://docs.travis-ci.com/user/sauce-connect/), I think we can largely take whether or not connect is running for granted.

For making sure connect is launched locally, I think that's a bit more challenging due to the number of ways it can be launched (command line, service, Firefox addon). I'm more a casual Unix user than a knowledgeable sysadmin, but short of grepping the process snapshot for Sauce Connect or the routing tables for what's using port 4445, I'm not sure if there's much we can do.

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgP0RaLwAHaX1z-w2Yd:b3fj97x) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Hmm, so if you want to use Firefox and geckodriver you need selenium 3.3.1. Is there a way to select latest selenium and set that for everything apart from Edge?
</blockquote></details>

I could do the inverse - set the default to `3.3.1` and override `if browserName == 'MicrosoftEdge'`. I chose 3.3.0 because that was the latest common version across [all the drivers](http://docs.seleniumhq.org/download/#client-drivers), but I have a feeling Sauce uses the Java driver, and would have 3.3.1.

---

*[wptrunner/environment.py, line 222 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJBmmzmpJTkzKzvR6E:-KgP7V3IPW7RcnhEDGIS:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/environment.py#L222)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

I think we shoud go back to a single environment and not add the complexity until we need it.
</blockquote></details>

Done.

---

*[wptrunner/executors/executorselenium.py, line 148 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJi8SZeu4vOl3e0CnK:-KgPD3mE6x827UOw_a8b:b2htzj5) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/executors/executorselenium.py#L148)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

My original patch was largely written whilst waiting to go into the theatre for a recording of the BBC Radio comedy panel show Just A Minute. It is entirely possible it does insane things. I suggest we try reverting this and see if there was any reason for it.
</blockquote></details>

I went ahead and rolled back all of the changes in this file. If it works for current `master`, it shouldn't need any changes to implement the Sauce feature (and I've [confirmed it still works](https://travis-ci.org/bobholt/web-platform-tests/builds/216340441)).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgUQTOPtMfXxsMvADHV:b-ui8nbc) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

I will add that there is a "Sauce Storage" feature that would allow uploading these scripts to the Sauce virtual machine via the Sauce REST API at build time and (it seems) setting the executable URI to something like `"sauce-storage:edge-prerun.bat"`I was hesitant to do that at first because of the added complexity (just trying to get it to work at all) and the addition of another potential point of failure. However, this approach would avoid having to refer to a URL like https://raw.githubusercontent.com/w3c/wptrunner/045487371a47581e59feafbae0bea78210aa3369/wptrunner/browsers/sauce_setup/edge-prerun.bat
</blockquote></details>

Follow up: I have a commit with the Sauce Storage solution working. In it, the `check_stability` script in web-platform-tests uploads the files before invoking wptrunner. The only change here is to set the URL to the sauce-storage URL. What that means is that the files wptrunner depends on to run sauce tests are uploaded to Sauce outside of wptrunner, which makes it a manual step if someone wants to run sauce tests locally. I don't see a way to get wptrunner to do the upload without breaking sauce concerns into the broader wptrunner context (the files have to exist before executing the test runner, and uploading them require sauce username and access key).

I'm okay with this, given that a user would have to set up Sauce Connect (a much more arduous manual process) in order to run sauce tests from a local wptrunner installation anyway. Uploading these files to Sauce storage would be one small additional step. However, I recognize that it's not ideal, and would have to be documented.

The other solution is to merge this as is, get the files into the master branch, and then make a quick enhancement pointing to those files rather than my gists or those files on my branch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgVLZvstNduec2whYTe:b-5f6o7z) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

Follow up: I have a commit with the Sauce Storage solution working. In it, the `check_stability` script in web-platform-tests uploads the files before invoking wptrunner. The only change here is to set the URL to the sauce-storage URL. What that means is that the files wptrunner depends on to run sauce tests are uploaded to Sauce outside of wptrunner, which makes it a manual step if someone wants to run sauce tests locally. I don't see a way to get wptrunner to do the upload without breaking sauce concerns into the broader wptrunner context (the files have to exist before executing the test runner, and uploading them require sauce username and access key).

I'm okay with this, given that a user would have to set up Sauce Connect (a much more arduous manual process) in order to run sauce tests from a local wptrunner installation anyway. Uploading these files to Sauce storage would be one small additional step. However, I recognize that it's not ideal, and would have to be documented.

The other solution is to merge this as is, get the files into the master branch, and then make a quick enhancement pointing to those files rather than my gists or those files on my branch.
</blockquote></details>

Can't wptrunner itself do the upload here? Is there some issue with just doing it on every run?

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgVLkiccPaU3K1IiKTP:b-oazaeb) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

Ah, I think I see what you're asking now. If I'm still being dense, feel free to talk it out with me on IRC.

I've done all of this through Travis, not setting up my own Sauce Connect locally. When we use this [through Travis](https://docs.travis-ci.com/user/sauce-connect/), I think we can largely take whether or not connect is running for granted.

For making sure connect is launched locally, I think that's a bit more challenging due to the number of ways it can be launched (command line, service, Firefox addon). I'm more a casual Unix user than a knowledgeable sysadmin, but short of grepping the process snapshot for Sauce Connect or the routing tables for what's using port 4445, I'm not sure if there's much we can do.
</blockquote></details>

Hmm, I meant "launch it from the command line" Like have a `--sauce-connect-binary` option and ensure that the binary is started at the start of the testrun and killed at the end, or assume it's running if that option isn't supplied. But I haven't ever used sauce connect so maybe there's something that would make that approach difficult?

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgVMJ6pID4o7Td5bP5r:bxhqc0i) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

I could do the inverse - set the default to `3.3.1` and override `if browserName == 'MicrosoftEdge'`. I chose 3.3.0 because that was the latest common version across [all the drivers](http://docs.seleniumhq.org/download/#client-drivers), but I have a feeling Sauce uses the Java driver, and would have 3.3.1.
</blockquote></details>

Yeah, I think the latter makes more sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgZ3MdW_y9gMYuiTbbE:b6pgyw3) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Can't wptrunner itself do the upload here? Is there some issue with just doing it on every run?
</blockquote></details>

It was lost in my rambling above, but my only issue is violating my perceived separation of concerns. The upload can't seem to happen in the `wptrunner/browsers/sauce.py`, but would need to happen earlier in `wptrunner/wptrunner.py`, itself. I'm not a fan of putting sauce-specific logic in there, but I suppose if I do it as a conditional on the sauce arguments getting passed in, that it's fine. 

I'll try to get that to work and add a commit that does it so you can see that change in isolation.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgZ7RqBEqWL8fiLhOjj:bwtzcmu) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Hmm, I meant "launch it from the command line" Like have a `--sauce-connect-binary` option and ensure that the binary is started at the start of the testrun and killed at the end, or assume it's running if that option isn't supplied. But I haven't ever used sauce connect so maybe there's something that would make that approach difficult?
</blockquote></details>

Ah, I see. I think that this is posslble. A simple sauce-connect setup should just need the binary, user name, and access key to execute.

Part of me wants this to be an enhancement after this PR lands so that I can do some additional testing, etc., but I'll get to work on it after uploading the pre-run scripts in wptrunner and see how it goes.

---

*[wptrunner/browsers/sauce.py, line 59 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJS5WmA8A717OEhXo7:-KgZCWzQjOCIC57o7xVC:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L59)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Yeah, I think the latter makes more sense.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 31 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJOzxCCYbLciMEj8DM:-KgZRpcRJucwtBQCWFRY:b-zayf0f) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L31)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

It was lost in my rambling above, but my only issue is violating my perceived separation of concerns. The upload can't seem to happen in the `wptrunner/browsers/sauce.py`, but would need to happen earlier in `wptrunner/wptrunner.py`, itself. I'm not a fan of putting sauce-specific logic in there, but I suppose if I do it as a conditional on the sauce arguments getting passed in, that it's fine. 

I'll try to get that to work and add a commit that does it so you can see that change in isolation.
</blockquote></details>

Done. Figured out a way to do it that doesn't leak Sauce-specific stuff to wptrunner.py (i.e. I realized I could just use the existing encapsulation scheme).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 5 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-Kg_2OLkBPiYhujYXc2i:b-mnzksd) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

Ah, I see. I think that this is posslble. A simple sauce-connect setup should just need the binary, user name, and access key to execute.

Part of me wants this to be an enhancement after this PR lands so that I can do some additional testing, etc., but I'll get to work on it after uploading the pre-run scripts in wptrunner and see how it goes.
</blockquote></details>

I got Sauce Connect running locally to see how the whole thing works. It should be possible to auto-launch, but doing so is a bit beyond my ken. Here's a breakdown of what needs to happen:

1. Invoke the Sauce Connect binary. This starts a long-running process that does not exit until directed.
2. "Halt" execution of wptrunner until...
3. Listen for the magic string "Sauce Connect is up, you may start your tests." in that process's STDOUT (this takes about 20 seconds on my machine).
4. When that string is received, execute the test runner.
5. When the tests are complete, exit the Sauce Connect process.

Like I said, this is a bit beyond my Python knowledge at the moment, so I'm happy to take some pointers at docs or sample implementations. I estimate it will take a bit of time for me to get it up and running, and would love to do it as an enhancement PR after this lands.

To make this PR more palatable in the interim, I have added (with an idea from @jugglinmike)  an error that gets thrown if it cannot connect to SauceConnect even before it tries to invoke webdriver: `Unable to connect to Sauce Labs. Is Sauce Connect Proxy running?` I've verified that it gets thrown locally if I don't have the tunnel open, is not thrown if it is open, and is not thrown on Travis.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

jugglinmike commented on wptrunner/browsers/sauce.py:
```diff
@@ -113,6 +112,13 @@ def env_options():
+    try:
+        tunnel_request = requests.get(url)
+    except requests.RequestException as e:
+        raise SauceException("Unable to connect to Sauce Labs. Is Sauce Connect Proxy running?")
```
Very nice!

jgraham commented:
Reviewed 5 of 13 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/products.py, line 46 at r5](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMNevkrfskwrGitBL:-KgoMNevkrfskwrGitBM:b-ofgez8) ([raw file](https://github.com/w3c/wptrunner/blob/8cb70a8465aab5402ee7fce1a0f865ac30aaf25d/wptrunner/products.py#L46)):*
> ```Python
>     executor_kwargs = getattr(module, data["executor_kwargs"])
>     env_options = getattr(module, data["env_options"])()
>     prerun = getattr(module, data.get("prerun") or "prerun", None)
> ```

Presumably this is just `data.get("prerun", "prerun")`

---

*[wptrunner/wptcommandline.py, line 198 at r3](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMDdI-WWUycofoQkx:-KgoMDdI-WWUycofoQky:bja6p0k) ([raw file](https://github.com/w3c/wptrunner/blob/bb19ef036b598fd8634af778acae77889d49448f/wptrunner/wptcommandline.py#L198)):*
> ```Python
>                              help="Sauce Connect tunnel identifier")
>     sauce_group.add_argument("--sauce-user", dest="sauce_user",
>                              default=os.environ.get("SAUCE_USERNAME"),
> ```

Why did you remove this?

---

*[wptrunner/wptrunner.py, line 127 at r5](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMXqngm6rnuAUw45y:-KgoMXqngm6rnuAUw45z:bpyzxje) ([raw file](https://github.com/w3c/wptrunner/blob/8cb70a8465aab5402ee7fce1a0f865ac30aaf25d/wptrunner/wptrunner.py#L127)):*
> ```Python
>         check_args(**kwargs)
> 
>         if prerun:
> ```

Ironically this should probably go in the environment class somehow.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgoLyGM8bmmDxHkiWyb:bxsyb6o) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

I got Sauce Connect running locally to see how the whole thing works. It should be possible to auto-launch, but doing so is a bit beyond my ken. Here's a breakdown of what needs to happen:

1. Invoke the Sauce Connect binary. This starts a long-running process that does not exit until directed.
2. "Halt" execution of wptrunner until...
3. Listen for the magic string "Sauce Connect is up, you may start your tests." in that process's STDOUT (this takes about 20 seconds on my machine).
4. When that string is received, execute the test runner.
5. When the tests are complete, exit the Sauce Connect process.

Like I said, this is a bit beyond my Python knowledge at the moment, so I'm happy to take some pointers at docs or sample implementations. I estimate it will take a bit of time for me to get it up and running, and would love to do it as an enhancement PR after this lands.

To make this PR more palatable in the interim, I have added (with an idea from @jugglinmike)  an error that gets thrown if it cannot connect to SauceConnect even before it tries to invoke webdriver: `Unable to connect to Sauce Labs. Is Sauce Connect Proxy running?` I've verified that it gets thrown locally if I don't have the tunnel open, is not thrown if it is open, and is not thrown on Travis.
</blockquote></details>

OK, this is fine for now. Please file an issue to do the followup work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[wptrunner/products.py, line 46 at r5](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMNevkrfskwrGitBL:-KgoabJnfe_l7vltqCbk:bvxxwlh) ([raw file](https://github.com/w3c/wptrunner/blob/8cb70a8465aab5402ee7fce1a0f865ac30aaf25d/wptrunner/products.py#L46)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Presumably this is just `data.get("prerun", "prerun")`
</blockquote></details>

Done in fixup commit.

---

*[wptrunner/wptcommandline.py, line 198 at r3](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMDdI-WWUycofoQkx:-KgoPeUh8KwTpQTKkcGa:bvbddf4) ([raw file](https://github.com/w3c/wptrunner/blob/bb19ef036b598fd8634af778acae77889d49448f/wptrunner/wptcommandline.py#L198)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Why did you remove this?
</blockquote></details>

All of the other defaults read from `os.environ` are in the check_stability script. It didn't make sense to break that convention and read from `os.environ` here for only these two variables.

---

*[wptrunner/wptrunner.py, line 127 at r5](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMXqngm6rnuAUw45y:-KgoaT2LQDsUK4wgRVLO:bdm9lqi) ([raw file](https://github.com/w3c/wptrunner/blob/8cb70a8465aab5402ee7fce1a0f865ac30aaf25d/wptrunner/wptrunner.py#L127)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Ironically this should probably go in the environment class somehow.
</blockquote></details>

Done as a new fixup commit. If there's a better way to do this, then I'm up for learning.

---

*[wptrunner/browsers/sauce.py, line 53 at r2](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgJUuhsOGjVtfU1bH6K:-KgoR8hSlxHk3m0SRpZO:bhyhzn1) ([raw file](https://github.com/w3c/wptrunner/blob/6f598ecc091be106e48b90d959149a20754f7e18/wptrunner/browsers/sauce.py#L53)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

OK, this is fine for now. Please file an issue to do the followup work.
</blockquote></details>

Will-do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[wptrunner/wptcommandline.py, line 198 at r3](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMDdI-WWUycofoQkx:-KgobLt2LhGqHJ62qpWx:b-cm7opc) ([raw file](https://github.com/w3c/wptrunner/blob/bb19ef036b598fd8634af778acae77889d49448f/wptrunner/wptcommandline.py#L198)):*
<details><summary><i>Previously, bobholt (Bob Holt) wrote…</i></summary><blockquote>

All of the other defaults read from `os.environ` are in the check_stability script. It didn't make sense to break that convention and read from `os.environ` here for only these two variables.
</blockquote></details>

So I would argue that a password is special and in that case at least it makes sense to support an environment variable directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

bobholt commented:
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[wptrunner/wptcommandline.py, line 198 at r3](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgoMDdI-WWUycofoQkx:-Kgof82esg3wh0NB3O0S:b-896fix) ([raw file](https://github.com/w3c/wptrunner/blob/bb19ef036b598fd8634af778acae77889d49448f/wptrunner/wptcommandline.py#L198)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

So I would argue that a password is special and in that case at least it makes sense to support an environment variable directly.
</blockquote></details>

Done.

---

*[wptrunner/browsers/sauce.py, line 120 at r5](https://reviewable.io:443/reviews/w3c/wptrunner/243#-KgFlF3dbuh0b7sHYBcy-r5-120:-KgofA5hoDxEDXjINpX4:bnr4fe1) ([raw file](https://github.com/w3c/wptrunner/blob/8cb70a8465aab5402ee7fce1a0f865ac30aaf25d/wptrunner/browsers/sauce.py#L120)):*
<details><summary><i>Previously, jugglinmike wrote…</i></summary><blockquote>

Very nice!
</blockquote></details>

Thanks, Mike!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/wptrunner/243)*
<!-- Sent from Reviewable.io -->
